### ISSUE TITLE
Starts getting 'nan' after aprx 290 iterations in modified Tiny_yolov3.

### ISSUE THREAD

**Issue by akshayjrkp** [AlexeyAB/darknet#979]
I am training a reduced Tiny_yolov3 (only single scale), for single class detetcion, without Batch Normailzation. Upto approximately 290 iterations the average loss reduces and comes close to 0, but then suddenly it shoots up and becomes 'nan' after 295 iterations.
If I change the same cfg fie with Batch Normalization = 1, then it trains fine.
What can be the issue? Please help.

Thanks in advance. Below is the cfg file.

```
[net]
# Testing
#batch=1
#subdivisions=1
# Training
batch=64
subdivisions=16
width=416
height=416
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.001
burn_in=1000
max_batches = 500200
policy=steps
steps=400000,450000
scales=.1,.1

[convolutional]
batch_normalize=0
filters=16
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=0
filters=32
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=0
filters=64
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=0
filters=128
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=0
filters=256
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=0
filters=512
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=1

[convolutional]
batch_normalize=0
filters=1024
size=3
stride=1
pad=1
activation=leaky

###########

[convolutional]
batch_normalize=0
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=0
filters=512
size=3
stride=1
pad=1
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=36
activation=linear



[yolo]
mask = 0,1,2,3,4,5
anchors = 31.7947,96.9781, 33.8494,112.6118, 38.3591,128.5938, 40.0406,150.9507, 55.5375,152.4082, 59.4220,190.0462
classes=1
num=6
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1

```

**Comment by AlexeyAB** [comment]
Try to train with `burn_in=0` and without Batch-normalization, may be it can help. And try to tweek learning rate.
But why do you want to train without batch-normalization?


https://arxiv.org/pdf/1502.03167v3.pdf

> 3.3 Batch Normalization enables higher
> learning rates
> In traditional deep networks, **too-high learning rate may
> result in the gradients that explode or vanish, as well as
> getting stuck in poor local minima. Batch Normalization
> helps address these issues**

...

> Batch Normalization achieves the same accuracy with 1**4 times fewer training steps**


https://arxiv.org/pdf/1612.08242v1.pdf

> Batch Normalization. Batch normalization leads to significant improvements in convergence while eliminating the need for other forms of regularization [7]. By adding batch normalization on all of the convolutional layers in YOLO we get more than 2% improvement in mAP. **Batch normalization also helps regularize the model. With batch normalization we can remove dropout** from the model without overfitting

**Comment by akshayjrkp** [comment]
I am targeting an FPGA solution for the tiny_yolo network. Thus if I train without BN, then I will not have to implement the BN layer for test time.
Is there any easy way in which I can remove BN on test time without removing it from the training?
I read and found that the previous conv. layer weights will need to be modified using the mena, sd, gamma and Beta values of the BN layer. But, that modification is not clear to me. Thus, I am planning to train without BN and then have the same weights used for test time on the FPGA.

I hope it makes sense. Please advice.
I will try to run with `burn_in=0` and report the result/error.

**Comment by AlexeyAB** [comment]
It is very bad idea to train without batch-normalization that is the best-practice for neural networks.
Try to read such explanation - how to fuse conv & batch-norm layers: http://machinethink.net/blog/object-detection-with-yolo/

You can just use such function to merge convolutional-layer **with** batch-norm --to--> convolutional-layer **without** batch-norm: https://github.com/AlexeyAB/darknet/blob/3010e1f755c67eeba03019727c40848ef9783799/src/detector.c#L1081

Implementation: https://github.com/AlexeyAB/darknet/blob/3010e1f755c67eeba03019727c40848ef9783799/src/network.c#L808-L844

----

> I hope it makes sense. Please advice.
> With burn_in=0 the avg loss becomes 'nan in the third iteration itself. It did not help, in any case can yo u explain what this parameter burn_in does?

burn_in=1000 means that until 1000 iterations the learning_rate will be equal to the = 
`lr * pow(iteration_number / 1000, 4)`: https://github.com/AlexeyAB/darknet/blob/3010e1f755c67eeba03019727c40848ef9783799/src/network.c#L94
